### DISCUSSION TITLE
Keeps getting a missing OAuth2 state parameter error on mobile (tested on iOS and Android)

### DISCUSSION THREAD

**Discussion by CreeperFarm** [pocketbase/pocketbase#6010] [Q&A] [answered]
I'm currently facing this error for a week a least, I know the error is not from my pocketbase configuration cause I tried to recreate a PocketBase DB to try the logging but it don't work, btw I also try on a clean new app with the example. I also have try to update my pocketbase and sdk, however, I'm still facing this error, do you have any idea why?

My Code : 
```dart

// Launch a page depending on platform
Future<void> _launchUrl(Uri url, context) async {
  debugPrint("Launching url : $url");
  try {
    await launchUrl(
      url,
      customTabsOptions: CustomTabsOptions(
        showTitle: true,
        shareState: CustomTabsShareState.on,
        urlBarHidingEnabled: true,
        closeButton: CustomTabsCloseButton(
          icon: CustomTabsCloseButtonIcons.back,
        ),
      ),
      safariVCOptions: SafariViewControllerOptions(
        preferredBarTintColor: Theme.of(context).colorScheme.surface,
        preferredControlTintColor: Theme.of(context).colorScheme.onSurface,
        barCollapsingEnabled: true,
        dismissButtonStyle: SafariViewControllerDismissButtonStyle.close,
      ),
    );
  } catch (e) {
    debugPrint(e.toString());
  }
}

signInWithGoogle(context) async {
  try {
    _pocketBase.authStore.clear();
    final authData = await _pocketBase.collection('users').authWithOAuth2('google', (url) async {
      await _launchUrl(url, context);
    }, scopes: [
      'https://www.googleapis.com/auth/userinfo.email',
      'https://www.googleapis.com/auth/userinfo.profile',
      //'https://www.googleapis.com/auth/user.gender.read',
      //'https://www.googleapis.com/auth/user.birthday.read',
    ], createData: {
      "email": "user${DateTime.now()}@mymangatheque.com",
      "username": "user${DateTime.now()}",
      "role": "user",
    });
    debugPrint(authData.toString());
    dynamic authData2 = await json.decode(authData.toString());

    var meta = authData.meta;
    if (_pocketBase.authStore.isValid) {
      if (authData2['meta']['isNew']) {
        var data = authData2['meta']['rawUser'];
        try {
          /*var imageId = await ImageDownloader.downloadImage(data['picture']);
      if (imageId == null) {
        return;
      }
      var fileName = await ImageDownloader.findName(imageId);
      var path = await ImageDownloader.findPath(imageId);*/
          var body = <String, dynamic>{
            "email": data['email'],
            //"birthday": DateTime.now().toString(),
          };

          // Upload the image of the user
          await _pocketBase.collection('users').update(
                _pocketBase.authStore.model.id, body: body, //files: [
                // Upload the image of the user
                /*http.MultipartFile.fromBytes(
          'avatar',
          File(path!).readAsBytesSync(),
          filename: fileName,
        )
          ]*/
              );

          await sendVerification(data['email']);
        } catch (e) {
          debugPrint(e.toString());
          showMessage("Un erreur est survenue.", context);
        }
      } else {
        debugPrint('User already exists');
      }
    } else {
      debugPrint('User isn\'t connected');
      showMessage('Une erreur est survenue', context);
    }
    _connectedUser.add(await findUser(authData2['meta']['rawUser']['email'].toString().toLowerCase()));
  } catch (e) {
    showMessage(e.toString(), context);
    debugPrint(e.toString());
  }
  await closeCustomTabs();
}

// The button that call the execution
// Google Button
SquareTile(
  imagePath: 'assets/images/google.png',
    onTap: () => {
    debugPrint("Google Sign In got clicked"),
    connector.signInWithGoogle(context),
    context.go('/profile'),
  },
),
```

**Comment by ganigeorgiev** [comment]
I can think of 2 possible cuprits:

1. Incorrectly configured OAuth2 app.

2. You are testing on Flutter Web and not on iOS/Android (see https://github.com/pocketbase/pocketbase/discussions/5874).
   If you are testing with an emulator using Flutter Web then the "all-in-one" OAuth2 call and the realtime events will not work because streamed responses are not supported on the web by `dart-lang/http` (_you can find a workaround under the [Limitations section](https://github.com/pocketbase/dart-sdk?tab=readme-ov-file#limitations)_).

The easiest way to rule out PocketBase issue is to test it with Dart cli using the file from the linked discussion:
```dart
// example.dart
import 'dart:io';

import 'package:pocketbase/pocketbase.dart';

void main() async {
  final pb = PocketBase('http://127.0.0.1:8090');

  final authData = await pb.collection('users').authWithOAuth2('google', (url) async {
    // calls a Linux shell command because `launchUrl` requires Flutter dependency
    // 
    // if you are testing on mac replace `xdg-open` with `open`
    Process.run('xdg-open', [url.toString()]);
  });

  print(authData);
}
```
_(run it with with `dart run example.dart`)_

If you are still not able to make it work, please create a **minimal** reproducible repo and I'll try to investigate it but based on my local tests it works correctly for me with both PocketBase v0.22.27 and v0.23.4.

**Reply by ganigeorgiev** [reply]
_Additionally, note that PocketBase Dart SDK v0.19+ is compatible with only PocketBase v0.23+ (see the [Dart SDK CHANGELOG](https://github.com/pocketbase/dart-sdk/blob/master/CHANGELOG.md))._

**Reply by ganigeorgiev** [reply]
_As anoter side-note, note that PocketBase v0.23 has builtin support for mapping common OAuth2 fields like name, avatar, username that can be configured from the collection options, so you might not need the extra update request:_
![oauth2_screenshot](https://camo.githubusercontent.com/36bf217f1e628bf72a12c306dd039257893356360c20af124c12ad3dbc4649d3/68747470733a2f2f692e696d6775722e636f6d2f365267386837342e706e67)

**Reply by CreeperFarm** [reply]
> I can think of 2 possible cuprits:
> 
> 1. Incorrectly configured OAuth2 app.
> 2. You are testing on Flutter Web and not on iOS/Android (see [[Dart] Problem with OAuth (Google and Facebook) #5874](https://github.com/pocketbase/pocketbase/discussions/5874)).
>    If you are testing with an emulator using Flutter Web then the "all-in-one" OAuth2 call and the realtime events will not work because streamed responses are not supported on the web by `dart-lang/http` (_you can find a workaround under the [Limitations section](https://github.com/pocketbase/dart-sdk?tab=readme-ov-file#limitations)_).
> 
> The easiest way to rule out PocketBase issue is to test it with Dart cli using the file from the linked discussion:
> 
> ```dart
> // example.dart
> import 'dart:io';
> 
> import 'package:pocketbase/pocketbase.dart';
> 
> void main() async {
>   final pb = PocketBase('http://127.0.0.1:8090');
> 
>   final authData = await pb.collection('users').authWithOAuth2('google', (url) async {
>     // calls a Linux shell command because `launchUrl` requires Flutter dependency
>     // 
>     // if you are testing on mac replace `xdg-open` with `open`
>     Process.run('xdg-open', [url.toString()]);
>   });
> 
>   print(authData);
> }
> ```
> 
> _(run it with with `dart run example.dart`)_
> 
> If you are still not able to make it work, please create a **minimal** reproducible repo and I'll try to investigate it but based on my local tests it works correctly for me with both PocketBase v0.22.27 and v0.23.4.

Like I said before, I'm using the mobile version and not the web, my dart sdk (0.19.1) and pocket (0.23.4) base are up to date and I tried the example but it doesn't work.

**Reply by ganigeorgiev** [reply]
If the Dart cli example doesn't work then I'm out of ideas. It works correctly for me when testing locally.

Without clear reproducible steps and details how did you create the OAuth2 app I don't think anyone will be able to assist you.

**Reply by CreeperFarm** [reply]
> cli example doesn't work then I'm ou

Maybe the error came from the fact that my oauth api is set as a prod's one ?

**Reply by ganigeorgiev** [reply]
I don't understand what do you mean by "oauth api is set as a prod's one".

As long as the configured redirect url host matches the one used in the SDK initialization, it should work fine. I've tested it again and reconfigured my local instance to use the Google provider for [app.presentator.io](https://app.presentator.io/#/login) and it works fine for me.

You can also start the executable with `--dev` and/or check the _Dashboard > Logs_ section for more details.

**Reply by CreeperFarm** [reply]
> instance to use the Google provider for [app.presentator.io](https://app.presentator.io/#/login) and it works fine for me.

The "oauth api is set as a prod's one" means that it's that it's open to anyone to connect as showed in the picture (this is on the OAuth Screen Consent Page).
<img width="343" alt="image" src="https://github.com/user-attachments/assets/ae6cec6e-790d-4175-99e8-b0dc47960013">

I'm thinking after this that maybe the error came from the fact that to show my api on web, it's using cloud flare tunnels so maybe it's blocking something.

**Reply by CreeperFarm** [reply]
Maybe it also can with my method to use the launchUrl.

**Reply by ganigeorgiev** [reply]
@CreeperFarm I have no energy to go through another back-and-forth and if users are not willing to follow the provided instructions then I cannot help further.

There are too many moving and unknown components in your setup to pinpoint the exact cause of the issue. The easiest way to rule out 3rd party misconfiguration is to try to test it locally with just PocketBase, no proxies, with the example as it is and to debug from there.

**Reply by CreeperFarm** [reply]
I'm really sorry I forgot to gave the component, do you want me ton send you the whole project through mail ?

**Comment by CreeperFarm** [comment, marked as answer]
After a few investigations, now it works, the reason of this error was because of Cloudflare caching even if deactivated on my API url so to make it works again, I have to let activated the developper mode.
